src/constraints.jl

```diff
@@ -0,0 +1,207 @@
+end
+
+if VERSION >= v"0.7-"
+    Base.broadcastable(cref::ConstraintRef) = Ref(cref)
+end
+
+"""
+    name(v::ConstraintRef)
+
+Get a constraint's name.
+"""
+name(cr::ConstraintRef{Model,<:MOICON}) = MOI.get(cr.m, MOI.ConstraintName(), cr)
+
+set_name(cr::ConstraintRef{Model,<:MOICON}, s::String) = MOI.set(cr.m, MOI.ConstraintName(), cr, s)
+
+"""
+    delete(model::Model, constraint_ref::ConstraintRef)
+
+Delete the constraint associated with `constraint_ref` from the model `model`.
+"""
+function delete(model::Model, constraint_ref::ConstraintRef{Model})
+    if model !== constraint_ref.m
+        error("The constraint reference you are trying to delete does not " *
+              "belong to the model.")
+    end
+    MOI.delete(model.moi_backend, index(constraint_ref))
+end
+
+"""
+    is_valid(model::Model, constraint_ref::ConstraintRef{Model})
+
+Return `true` if `constraint_ref` refers to a valid constraint in `model`.
+"""
+function is_valid(model::Model, constraint_ref::ConstraintRef{Model})
+    return (model === constraint_ref.m &&
+            MOI.is_valid(model.moi_backend, constraint_ref.index))
+end
+
+#############################################################################
+# AbstractConstraint
+# Abstract base type for all constraint types
+abstract type AbstractConstraint end
+
+struct ScalarConstraint{F <: AbstractJuMPScalar,
+                        S <: MOI.AbstractScalarSet} <: AbstractConstraint
+    func::F
+    set::S
+end
+
+moi_function_and_set(c::ScalarConstraint) = (moi_function(c.func), c.set)
+shape(::ScalarConstraint) = ScalarShape()
+function constraint_object(ref::ConstraintRef{Model, MOICON{FuncType, SetType}}) where
+        {FuncType <: MOI.AbstractScalarFunction, SetType <: MOI.AbstractScalarSet}
+    model = ref.m
+    f = MOI.get(model, MOI.ConstraintFunction(), ref)::FuncType
+    s = MOI.get(model, MOI.ConstraintSet(), ref)::SetType
+    return ScalarConstraint(jump_function(model, f), s)
+end
+
+struct VectorConstraint{F <: AbstractJuMPScalar,
+                        S <: MOI.AbstractVectorSet,
+                        Shape <: AbstractShape} <: AbstractConstraint
+    func::Vector{F}
+    set::S
+    shape::Shape
+end
+function VectorConstraint(func::Vector{<:AbstractJuMPScalar},
+                          set::MOI.AbstractVectorSet)
+    VectorConstraint(func, set, VectorShape())
+end
+
+moi_function_and_set(c::VectorConstraint) = (moi_function(c.func), c.set)
+shape(c::VectorConstraint) = c.shape
+function constraint_object(ref::ConstraintRef{Model, MOICON{FuncType, SetType}}) where
+        {FuncType <: MOI.AbstractVectorFunction, SetType <: MOI.AbstractVectorSet}
+    model = ref.m
+    f = MOI.get(model, MOI.ConstraintFunction(), ref)::FuncType
+    s = MOI.get(model, MOI.ConstraintSet(), ref)::SetType
+    return VectorConstraint(jump_function(model, f), s, ref.shape)
+end
+
+# TODO: Find somewhere to put this error message.
+#add_constraint(m::Model, c::Array{AffExprConstraint}) =
```

Comment:
Do you know what the intention of this code is? I don't know what error case it handles and why `add_constraint` would say something about macro syntax.

Reply:
We can remove it, it is done in
https://github.com/JuliaOpt/JuMP.jl/blob/672de56de2e7462dfe454a8b1b716eca40b74f19/src/macros.jl#L362